sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -888,6 +888,27 @@ pi_result _pi_queue::executeCommandList(ze_command_list_handle_t ZeCommandList,
     this->ZeOpenCommandListSize = 0;
   }
 
+  // We are going to submit kernels for execution. If indirect access flag is
+  // set for a kernel then we need to make a snapshot of existing memory
+  // allocations. We need to lock the mutex guarding the list of memory
+  // allocations in the platform to prevent creation of new memory alocations
+  // before we submit the kernel for execution.
+  // Note. Currently indirect access flag is set for all kernels to be
+  // conservative, that's why no need to check for a flag.
+  std::lock_guard<std::mutex> MemAllocsLock(Device->Platform->MemAllocsMutex);
```

Comment:
Since this is certainly run-time overhead, and may lead to out-of-memory issues, we certainly need to guard this with a flag (new env var, I guess). Also maybe have it OFF by default.

Reply:
I guarded changes of this commit.  I guarded zeKernelSetIndirectAccess with the same guard, because if we set this flag we need to support deferred release. Do you agree with this?

Reply:
No. If we do this, then there is no way for running the kernel correctly without the overhead introduced with this patch. I think we should provide an option to users that don't experience problem being fixed to run at full speed.

Reply:
I removed the guard from  zeKernelSetIndirectAccess, So it is always set as before.